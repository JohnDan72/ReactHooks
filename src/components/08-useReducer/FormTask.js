import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import { Button, ButtonToolbar, Form } from 'rsuite';
import { useSimpleForm } from '../../hooks/useSimpleForm';

const propTypes = {
    handleAdd: PropTypes.func.isRequired,
};
const defaultProps = {};


const FormTask = ({ handleAdd }) => {
    const inputRef = useRef();

    const { formValue, handleInputChange } = useSimpleForm({ task: '' });
    const { task } = formValue;

    const handleSumbit = (e) => {

        if (formValue.task.trim() === '') {
            console.log("Campo vac√≠o");
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: formValue.task,
            done: false
        }
        

        handleInputChange({ name: 'task', value: '' });
        // inputRef.current.firstChild.value = '';
        // inputRef.current.firstChild.select();
        handleAdd(newTodo);

    }
    return (
        <Form onSubmit={handleSumbit}>
            <Form.Group controlId="name">
                <Form.ControlLabel>Task description</Form.ControlLabel>
                <Form.Control name="task"
                    placeholder="new task"
                    ref={inputRef}
                    value={task}
                    onChange={(value) => handleInputChange({ name: 'task', value })}
                />

            </Form.Group>
            <Form.Group>
                <ButtonToolbar >
                    <Button appearance="primary" type="submit">Add Item</Button>
                </ButtonToolbar>
            </Form.Group>
        </Form>
    );
}

FormTask.propTypes = propTypes;
FormTask.defaultProps = defaultProps;

export default FormTask;