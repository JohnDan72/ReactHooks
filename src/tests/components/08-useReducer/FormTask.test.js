const { shallow } = require("enzyme")
import FormTask from "../../../components/08-useReducer/FormTask";

describe('Pruebas FormTask', () => {
    
    const handleAdd = jest.fn();

    test('debe renderizar el component', () => {
        const wrapper = shallow(<FormTask handleAdd={handleAdd}/>);
        // console.log(wrapper.props())
        expect(wrapper).toMatchSnapshot();
    });
    
    test('no debe llamar handleAdd todo', () => {
        const wrapper = shallow(<FormTask handleAdd={handleAdd}/>);
        const formsubmit = wrapper.find('Form').prop('onSubmit');
        formsubmit();

        expect(handleAdd).toHaveBeenCalledTimes(0);

    });

    test('debe llamar handleAdd todo con un argumento', async() => {
        const wrapper = shallow(<FormTask handleAdd={handleAdd}/>);
        const value = 'Aprender Python';
        wrapper.find('FormControl').simulate('change',value);
        
        // NOTA: llamar a esta función despues de la simulación, evita null inesperados
        const formsubmit = wrapper.find('Form').prop('onSubmit');
        formsubmit();

        expect(handleAdd).toHaveBeenCalledTimes(1);
        expect(handleAdd).toHaveBeenCalledWith( expect.any(Object));
        expect(handleAdd).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false
        });

    });
})