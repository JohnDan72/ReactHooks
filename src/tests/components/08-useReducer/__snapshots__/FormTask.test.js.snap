// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas FormTask debe renderizar el component 1`] = `
<Form
  onSubmit={[Function]}
>
  <FormGroup
    controlId="name"
  >
    <FormControlLabel>
      Task description
    </FormControlLabel>
    <FormControl
      name="task"
      onChange={[Function]}
      placeholder="new task"
      value=""
    />
  </FormGroup>
  <FormGroup>
    <ButtonToolbar>
      <Button
        appearance="primary"
        type="submit"
      >
        Add Item
      </Button>
    </ButtonToolbar>
  </FormGroup>
</Form>
`;
